File: codigo/matlab/error_relativo.m
% error_relativo.m -- Errores absoluto, relativo y aproximado. Chapra cap. 3.
% Script con funcion local (requiere R2016b+ / Octave).

[ea, er] = errores(pi, 22/7);
fprintf('Aproximacion 22/7 de pi:\n');
fprintf('  Error absoluto = %.6e\n', ea);
fprintf('  Error relativo = %.6e (%.4f%%)\n', er, er*100);

% Criterio de paro por error relativo aproximado entre iteraciones
tol = 1e-4; x_prev = 1; x_curr = 1.5; k = 0;
while true
    k = k + 1;
    erp = abs((x_curr - x_prev)/x_curr);      % error relativo aproximado
    if erp < tol, break; end
    x_prev = x_curr;
    x_curr = 0.5*(x_curr + 2/x_curr);         % iteracion (raiz de 2)
end
fprintf('\nRaiz de 2 ~= %.8f en %d iteraciones (erp < %.0e)\n', x_curr, k, tol);

function [ea, er] = errores(x_real, x_aprox)
    ea = abs(x_aprox - x_real);
    er = ea / abs(x_real);
end

File: codigo/python/error_relativo.py
"""error_relativo.py -- Errores absoluto, relativo y aproximado. Chapra cap. 3."""
import math


def errores(x_real, x_aprox):
    """Devuelve (error absoluto, error relativo)."""
    ea = abs(x_aprox - x_real)
    er = ea / abs(x_real)
    return ea, er


if __name__ == "__main__":
    ea, er = errores(math.pi, 22 / 7)
    print("Aproximacion 22/7 de pi:")
    print(f"  Error absoluto = {ea:.6e}")
    print(f"  Error relativo = {er:.6e} ({er * 100:.4f}%)")

    # Criterio de paro por error relativo aproximado entre iteraciones
    tol, x_prev, x_curr, k = 1e-4, 1.0, 1.5, 0
    while True:
        k += 1
        erp = abs((x_curr - x_prev) / x_curr)   # error relativo aproximado
        if erp < tol:
            break
        x_prev, x_curr = x_curr, 0.5 * (x_curr + 2 / x_curr)
    print(f"\nRaiz de 2 ~= {x_curr:.8f} en {k} iteraciones (erp < {tol:.0e})")
